Shopify/ruby-lsp
# Return error response when exceptions are raised

### Motivation

LSP [supports `error` attribute in response](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseMessage), which will print the error in the extension output **even if tracing is not enabled**. I think this will help use debug issues a lot. 

**Example**

```
[Error - 2:48:24 PM] Request textDocument/formatting failed.
  Message: #<OptionParser::InvalidOption: --auto-foo>
  Code: -32603 
{:jsonrpc=>"2.0", :id=>17, :method=>"textDocument/formatting", :params=>{:textDocument=>{:uri=>"file:///Users/hung-wulo/src/github.com/Shopify/ruby-lsp/lib/ruby_lsp/requests/formatting.rb"}, :options=>{:tabSize=>2, :insertSpaces=>true, :trimTrailingWhitespace=>true, :insertFinalNewline=>true}}}
```

### Implementation

When there's an exception inside `Handler#handle`, we write back the response with `error` attribute instead of `result`.

### Automated Tests

Based on the response to my question in https://github.com/Shopify/ruby-lsp/pull/149 (none), I think we're ok with not covering error handling with a new category of tests for now. 

### Manual Tests

1. Change the `--auto-correct` flag in `formatting.rb` to something like `--foo`
2. Restart LSP
3. Open a Ruby file, mess up indentations, and save
4. The error should show up in the `Ruby LSP` output and look like

```
[Error - 2:48:24 PM] Request textDocument/formatting failed.
  Message: #<OptionParser::InvalidOption: --auto-foo>
  Code: -32603 
{:jsonrpc=>"2.0", :id=>17, :method=>"textDocument/formatting", :params=>{:textDocument=>{:uri=>"file:///Users/hung-wulo/src/github.com/Shopify/ruby-lsp/lib/ruby_lsp/requests/formatting.rb"}, :options=>{:tabSize=>2, :insertSpaces=>true, :trimTrailingWhitespace=>true, :insertFinalNewline=>true}}}
```
